src/ImageSharp/Memory/Buffer2D{T}.cs

```diff
@@ -62,17 +71,76 @@ internal Buffer2D(MemorySource<T> memorySource, int width, int height)
                 DebugGuard.MustBeLessThan(x, this.Width, nameof(x));
                 DebugGuard.MustBeLessThan(y, this.Height, nameof(y));
 
-                Span<T> span = this.GetSpan();
-                return ref span[(this.Width * y) + x];
+                return ref this.GetRowSpanUnchecked(y)[x];
             }
         }
 
+        /// <summary>
+        /// Gets a <see cref="Span{T}"/> to the row 'y' beginning from the pixel at the first pixel on that row.
+        /// </summary>
+        /// <param name="y">The y (row) coordinate.</param>
+        /// <returns>The <see cref="Span{T}"/>.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Span<T> GetRowSpan(int y)
+        {
+            Guard.MustBeGreaterThanOrEqualTo(y, 0, nameof(y));
+            Guard.MustBeLessThan(y, this.Height, nameof(y));
+
+            return this.cachedMemory.Length > 0
+                ? this.cachedMemory.Span.Slice(y * this.Width, this.Width)
+                : this.GetRowMemorySlow(y).Span;
+        }
+
         /// <summary>
         /// Disposes the <see cref="Buffer2D{T}"/> instance
         /// </summary>
         public void Dispose()
         {
-            this.MemorySource.Dispose();
+            this.MemoryGroup.Dispose();
+            this.cachedMemory = default;
+        }
+
+        /// <summary>
+        /// Same as <see cref="GetRowSpan"/>, but does not validate index in Release mode.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Span<T> GetRowSpanUnchecked(int y)
```

Comment:
I feel like this is something that could be handled better with compiler conditionals rather than two separate methods. It's a little confusing knowing which to call although it does seem the other version is only called via tests.

Reply:
I'm also unsure about this solution. Normally, the checked variant should be good and fast enough, and that is the only public variant, so ImageSharp.Drawing will not be able to rely on the unchecked one.

On the other hand, there are some special cases like Jpeg where we need the extra perf from omitting the bouds checking:
https://github.com/SixLabors/ImageSharp/blob/9c6f1180393145dcc8009b9d563d36ccb42fdaf0/src/ImageSharp/Formats/Jpeg/Components/RowOctet.cs#L30-L37

Reply:
I see the dilemma, ok, since it's not public lets carry on as is.

Reply:
@antonfirsov I've been rethinking this and I think we should use the unchecked version as the public version. My reasoning is that we do not do any explicit bounds checks for `Image<TPixel>[x,y]` in release mode and rely on the framewrok check/massages instead.

Reply:
The problem with this is that it goes against [design guidelines for public API exception safety](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/using-standard-exception-types#nullreferenceexception-indexoutofrangeexception-and-accessviolationexception).

It's not only about bounds-checking: we also need to decide what to do with `null` safety, doesn't make sense to deal with only one of thise concerns. (Guideline suggestion is not to leak `NullReferenceException`!)

Of course we may decide to compromise on this (pro: perf, thinner API surface), but we probably don't want to. It would be better IMO to eventually expose the unsafe/unchecked variants to advanced users / downstream libs, since their dangerous nature is obvious from the naming.

Reply:
Is this something we should fix within the encoder itself. The idea of two methods that do the same thing with and without checking irks me.

Reply:
Not with the current design. There is a private member, `cachedMemory`, to speed up serving requests for the most common single-buffer case:
https://github.com/SixLabors/ImageSharp/blob/b35e3cbfbd133eb0d3290a2d537d40f649444133/src/ImageSharp/Memory/Buffer2D%7BT%7D.cs#L112-L114

Even if could manage to solve this (which I doubt we can do in an elegant way), should we duplicate the unchecked code everywhere across the lib? A few places where we definitely need it:
- `image[x,y]` indexer (to avoid doing bounds-checking twice)
- `DenseMatrixUtils.Convolve2DImpl()`
- Resize
- Jpeg encoder & decoder

I can't see a win-all situation here we either:
1. Compromise performance
2. Compromise public API exception guidelines
3. Keep the unchecked variants for internal/advanced scenarios

For me the 3rd is the lesser evil.

Note that I created the two variants by renaming the old `.GetRowSpan(y)` to `.GetRowSpanUnchecked(y)`, and then creating a new `GetRowSpan(y)` that is actually checked, to make sure I don't regress performance. The original `.GetRowSpan(y)` was unchecked and public 😕

Reply:
Rethinking it again, actually ignoring the guideline on *advanced types*, might be also an option.

Most important is to make a conscious decision, so far `Buffer2D` was non-conformant by accident.

Reply:
I guess for now we have to stick with option 3 then and ensure the method docs indicate its use case so we don't abuse it.

Reply:
I'm so good at changing my mind in just a few minutes, I already started coding option 2 right now 😄

Should I stop?

Reply:
Hahaha....  Well.... They are guidelines not explicit rules. 

My thinking is this. Those exceptions are "reserved" for the framework but by relying on them instead of relying on our own checks we do not expose the user to any greater risk unless we use unsafe code without bounds checks. You and I both agree that we should severely limit usage of the unsafe code so I would like to say to be damned with the guidelines. The fact we are having this conversation suggests we know what we are doing.

Reply:
Those guidelines are here to return useful error information for typical users, saving time for everyone.

But this also indicates, that we can make exceptions for (and only for!) advanced scenarios, if we properly document them. Most people won't even notice that `Buffer2D<T>` exists.

Reply:
Pushed changes, including the update of the public indexers, it turned out to be much better than I expected.

Reply:
Excellent. Will have a look tonight.